Ockam command transports API

spacekookie opened the pull request:
Also add a `transport` command to the ockam command to create and
manage transports for nodes.

<!-- Thank you for sending a pull request :heart: -->

## Current Behaviour

<!-- Please describe the current behavior of the code before the changes in this pull request is applied. -->

## Proposed Changes

<!-- Please describe the changes proposed in this pull request. -->
<!-- If this pull request resolves an already recorded bug or a feature request, please add a link to that issue. -->

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [ ] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- [ ] All commits in this Pull Request follow the Ockam [commit message convention](https://www.ockam.io/learn/how-to-guides/contributing/CONTRIBUTING#commit-messages).
- [ ] I accept the Ockam Community [Code of Conduct](https://www.ockam.io/learn/how-to-guides/high-performance-team/conduct/).
- [ ] I have accepted the Ockam [Contributor License Agreement](https://www.ockam.io/learn/how-to-guides/contributing/cla/) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/contributors](https://github.com/build-trust/contributors) repository.

<!-- Looking forward to merging your contribution!! -->

spacekookie committed:
refactor(rust): move nodeman protocol definitions to submodule

Also add a `transport` command to the ockam command to create and
manage transports for nodes.

spacekookie committed:
refactor(rust): update nodemanager service to ockam_api structures

spacekookie committed:
feat(rust): add configuration management to ockam_command

spacekookie committed:
fix(rust): improve the usability of ockam command

spacekookie committed:
refactor(rust): move node manager service to ockam_api crate

twittner commented on implementations/rust/ockam/ockam_api/Cargo.toml:
```diff
@@ -25,6 +25,7 @@ lmdb          = ["std", "lmdb-rkv"]
 vault-storage = ["ockam_vault/storage"]
 
 [dependencies]
+ockam       = { path = "../ockam", version = "^0.60.0", features = ["software_vault"] }
```
```suggestion
ockam       = { path = "../ockam", version = "0.60.0", features = ["software_vault"] }
```

mrinalwadhwa replied:
Our release scripts manage/update versions at release time within our crates @metaclips may have a recommendation based on what the scripts expect.

twittner commented on implementations/rust/ockam/ockam_api/src/lib.rs:
```diff
@@ -11,8 +11,12 @@ use minicbor::encode::{self, Encoder, Write};
 use minicbor::{Decode, Encode};
 use ockam_core::compat::borrow::Cow;
 use ockam_core::compat::rand;
+use std::fmt::{Display, Formatter};
```
```suggestion
use core::fmt::{Display, Formatter};
```

twittner commented on implementations/rust/ockam/ockam_api/src/lib.rs:
```diff
@@ -89,6 +93,22 @@ pub enum Method {
+                Self::Delete => "DELETE",
+                Self::Patch => "PATCH",
+            }
+        )
```
```suggestion
        f.write_str(
            match self {
                Self::Get => "GET",
                Self::Post => "POST",
                Self::Put => "PUT",
                Self::Delete => "DELETE",
                Self::Patch => "PATCH",
            }
        )
```

twittner commented on implementations/rust/ockam/ockam_api/src/nodes/types.rs:
```diff
@@ -1,106 +1,156 @@
-    #[cfg(feature = "tag")]
-    #[n(0)] tag: TypeTag<4925167>,
-    #[b(1)] name: Cow<'a, str>
+pub enum TransportType {
```
Just wanted to mention that if an enum will never carry data in its variants, one may consider using [`#[cbor(index_only)]`](https://twittner.gitlab.io/minicbor/minicbor_derive/index.html#cborindex_only).

twittner commented on implementations/rust/ockam/ockam_api/src/nodes/types.rs:
```diff
@@ -1,106 +1,156 @@
+                Self::Ble => "BLE",
+                Self::WebSocket => "Websocket",
+            }
+        )
```
```suggestion
        f.write_str(
            match self {
                Self::Tcp => "TCP",
                Self::Ble => "BLE",
                Self::WebSocket => "Websocket",
            }
        )
```

twittner commented on implementations/rust/ockam/ockam_api/src/nodes/types.rs:
```diff
@@ -1,106 +1,156 @@
+                Self::Listen => "Listening",
+                Self::Connect => "Remote connection",
+            }
+        )
```
```suggestion
        f.write_str(
            match self {
                Self::Listen => "Listening",
                Self::Connect => "Remote connection",
            }
        )
```

twittner commented on implementations/rust/ockam/ockam_api/src/nodes/types.rs:
```diff
@@ -1,106 +1,156 @@
             tag: TypeTag,
-            name: name.into(),
+            node_name: Cow::Borrowed(node_name),
+            status: Cow::Owned(status.into()),
```
Any reason why status has to be owned here?

twittner commented on implementations/rust/ockam/ockam_api/src/nodes/types.rs:
```diff
@@ -1,106 +1,156 @@
-    pub fn status(&self) -> &str {
-        &self.status
+impl<'a> TransportStatus<'a> {
+    pub fn new(tt: TransportType, tm: TransportMode, addr: &String) -> Self {
+        Self {
+            #[cfg(feature = "tag")]
+            tag: TypeTag,
+            tt,
+            tm,
+            addr: addr.clone().into(),
```
```suggestion
    pub fn new<S: Into<Cow<'a, str>>>(tt: TransportType, tm: TransportMode, addr: S) -> Self {
        Self {
            #[cfg(feature = "tag")]
            tag: TypeTag,
            tt,
            tm,
            addr: addr.into(),
```

spacekookie committed:
fix(rust): apply style feedback

spacekookie committed:
feat(rust): implement transport creation via ockam command

Add the ability to create new transports via the node manager
service.  Status about the creation will be relayed to the calling
CLI.  Any error will be displayed to the user, and new transports
tracked via the node manager.

Currently it's not yet possible to delete a transport, but the route
to the newly created transport is displayed.

spacekookie committed:
feat(rust): allow indexing of transports via unique ids

spacekookie committed:
feat(rust): support deleting existing transports

spacekookie committed:
feat(rust): basic inlet and outlet messaging request bodies
